documentation/mqtt-over-amqp/docs/amqp_qos_retain.md

```diff
@@ -0,0 +1,94 @@
+
+The following document describes a proposal for defining the meaning of the QoS level and retain flag on messages moving
+from the MQTT to AMQP specification. This concepts are already used today in the MQTT gateway and LWT services for EnMasse
+but they could be defined in a more general way as not related at MQTT but just at AMQP level, because we want to have a
+more generic AMQP LWT service (or "Connectivity" service).
+ 
+## QoS Level
+
+## What happens in the MQTT land
+
+With the MQTT protocol you can define an end-to-end delivery guarantee from the publisher to the subscriber through a broker.
+The QoS is specified at :
+
+* _subscription level_, when a subscriber subscribes to a topic asking for a desired QoS
+* _message level_, when the publisher sends the message with such QoS specified in the message itself
+
+At _subscription level_, the subscriber asks for a desired QoS on the topic and it'd like to receive messages on that topic published
+with such QoS. As reply to the subscription request, the broker provides the "granted" QoS for the topic which could be
+the same asked by the subscriber or lower than that (i.e. subscriber wants QoS 1 on topic "foo" but broker replies with
+granted QoS 0 on topic "foo". Here the subscriber can decide to close the connection if it needs QoS 1).
+
+At _message level_, it means that the publisher sends a message specifying the QoS inside it and this QoS drives the
+acknowledgment steps with the broker; so it's just about the "contract" between publisher and broker.
+
+For having an end-to-end delivery guarantee, both the QoS levels play their rule.
+When the broker has to re-publish a message published with QoS X to a subscriber which is subscribed with QoS Y, the following
+criteria is applied :
+
+`re-publish QoS = min(published QoS, granted QoS)`
+
+In this case the re-publish QoS will be the QoS specified at _message level_ from broker to subscriber which will drive
+the acknowledgment steps between them.
+
+## What it could mean in the AMQP land
+
+With the AMQP protocol the "settlement" is defined between the two peers (during link attachment with "modes" and on message with "settled") 
+and how publisher sends the message to a broker doesn't effect how this message flows from broker to receiver.
+
+> It's something different from the MQTT case where the re-publish QoS is influenced by how the publisher sends and how the 
+subscriber wants to receive.
+
+In AMQP the QoS could be specified at :
+
+* _link level_ , when the receiver attaches a link specifying the desired QoS
+* _message level_, when the sender sends the message with such QoS specified in the message itself
+
+At _link level_, we could have :
+
+* a new "desired-capabilities" named `qos-policy` during the link attachment from the receiver
+* a link "property" specifying the value for the above `qos-policy`
+* the source should support such "capability"
+* a link "property" specifying the value for the desired QoS on receiving
```

Comment:
My initial reaction here is that the AMQP spec already defines a way for peers to select a QoS for the link, through the sender- and receiver- settlement modes. While a sender-settlement mode of 'mixed' indicates a per-message choice, there isn't a way - as far as I can tell - to decide on the receiver settlement mode per message. That means that the same link cannot be used for receiving both at-most-once/at-least-once aka Qos 0/1 and exactly-once aka QoS 2 messages.

Personally, I'm not sure I see sufficient value in the MQTT approach to make it worth trying to define something this complex. The primary value I see is that the receiver is made aware of the fact that a message may have been published with a QoS less than it itself negotiated, which may effect the end-to-end delivery guarantee.

Perhaps a simpler solution would just be to provide some way of indicating the QoS/delivery guarantee under which the original publication was made. The receiver would use the same pattern for its own acknowledgement of all messages, but would be able to determine where the expectations between itself and the publisher differ.

Reply:
@gemmellr I realise it has been about a month since you left this comment, but for some reason I missed it until now :-)

To me the issue is that the way the attach definition is worded around "rcv-settle-mode": 

> When set at the receiver this indicates the actual settlement mode in use" in combination 

with the definition of the mode "second" : 

> The receiver will only settle after sending the disposition to the sender and receiving a disposition indicating settlement of the delivery from the sender.

Makes it appear that the receiver never has discretion to actually settle first.  As per your reference to the rcv-settle-mode on transfer, obviously here receiver is being instructed to settle first (even though the mode is second).  And, in general, I think either side can choose to settle a delivery at any point in time... therefore the receiver choosing rcv-settle-mode "second" is not precluding the receiver actually settling first where either it is instructed to by the sender, or because it has it's own reasons why it needs to settle immediately upon achieving a terminal state.  The wording could do with clarifying though,

Reply:
@rgodfrey On the point about what attach says, I believe thats just related to the fact the peer receiving the attach might reply with different value depending on what it supports (i.e something that doesnt support 'second' probably shouldn't respond with 'second' but could respond with 'first'), with the next sentence there essentially also covering that.

The definition for Transfer seems to me to say its value if set overrides the link attach value but is only useful at all for setting 'first' on a transfer on a link originally attached with 'second'. Thus you could use mixed+second to support sending at-most-once settled, exactly-once unsettled, and at least once unsettled transfers on the link.

I don't think the spec really allows for either side to settle at any point in general when 'second' is used, with the above instead being the mechanism to make particular messages use 'first', though it seems to me there are some obvious cases there would effectively be no choice for a receiver but to do so....so, yeah..